thavalon-server/src/game/state/on_mission.rs

```diff
@@ -0,0 +1,334 @@
+use std::collections::HashMap;
+use std::time::Duration;
+
+use super::prelude::*;
+
+/// Phase for when the voted-on players are going on a mission
+pub struct OnMission {
+    /// The proposal this mission is based on. Used to check if players are on the mission or not
+    proposal_index: usize,
+
+    /// Cards played by each player on the mission.
+    cards: HashMap<String, Card>,
+
+    /// The number of questing beasts played
+    questing_beasts: usize,
+}
+
+/// Placeholder phase used when waiting for Agravaine to declare
+pub struct WaitingForAgravaine {
+    proposal_index: usize,
+}
+
+/// How long to wait for an Agravaine declaration
+const AGRAVAINE_TIMEOUT: Duration = Duration::from_secs(30);
+
+impl GameState<OnMission> {
+    pub fn handle_card(mut self, player: &str, card: Card) -> ActionResult {
+        if self.includes_player(player) {
+            if let Some(card) = self.phase.cards.get(player).cloned() {
+                self.player_error(format!("You already played a {}", card))
+            } else if !self.game.players.by_name(player).unwrap().role.can_play(card) {
+                self.player_error(format!("You can't play a {}", card))
+            } else {
+                self.phase.cards.insert(player.to_string(), card);
+                log::debug!("{} played a {}", player, card);
+
+                if self.phase.cards.len() == self.proposal().players.len() {
+                    let mission = self.mission();
+                    let passed =
+                        !is_failure(self.game.spec, mission as usize, self.phase.cards.values());
+                    log::debug!(
+                        "Mission {} {}",
+                        mission,
+                        if passed { "passed" } else { "failed" }
+                    );
+
+                    self.mission_results.push(MissionResults {
+                        passed,
+                        players: self.proposal().players.clone(),
+                    });
+
+                    let (mut successes, mut fails, mut reverses) = (0, 0, 0);
+                    for card in self.phase.cards.values() {
+                        match card {
+                            Card::Success => successes += 1,
+                            Card::Fail => fails += 1,
+                            Card::Reverse => reverses += 1,
+                        }
+                    }
+
+                    let mut effects = vec![Effect::Broadcast(Message::MissionResults {
+                        mission,
+                        successes,
+                        fails,
+                        reverses,
+                        questing_beasts: self.phase.questing_beasts,
+                        passed,
+                    })];
+
+                    // TODO: how does Agravaine work on mission 4?
+                    if self.game.spec.has_role(Role::Agravaine) && passed {
```

Comment:
Not sure if `has_role` just checks if the role is possible or is actually in the game. If it's the latter, we should remove this check, as it leaks information that an Agravaine is in the game if the effect only fires if an Agravaine is present.

Reply:
This is `GameSpec::has_role`, which only checks if the role is possible. There's also a `Players::has_role`, which checks if it's in the game. Having a check here isn't strictly necessary, but it means that smaller games don't have to wait for a potential Agravaine declaration.

Reply:
In that case, I like it. Just wanted to make sure we weren't leaking information.